ASoC: Intel: sof_sdw: use generic rtd_init function for Realtek SDW DMICs

bardliao opened the pull request:
The only thing that the rt_xxx_rtd_init() functions do is to set
card->components. And we can set card->components with name_prefix
as rt712_sdca_dmic_rtd_init() does.
And sof_sdw_rtd_init() will always select the first dai with the
given dai->name from codec_info_list[]. Unfortunately, we have
different codecs with the same dai name. For example, dai name of
rt715 and rt715-sdca are both "rt715-aif2". Using a generic rtd_init
allow sof_sdw_rtd_init() run the rtd_init() callback from a similar
codec dai.

Fixes: https://github.com/thesofproject/linux/issues/4834

ujfalusi commented:
@bardliao, you cannot change the name_prefix as UCM expects `rt714 ` for sdca variants:
```
# cat /usr/share/alsa/ucm2/codecs/rt715-sdca/init.conf
# RT715-sdca (aka RT714) specific volume control settings

BootSequence [
        cset "name='rt714 FU0A Capture Switch' 0"
        cset "name='rt714 ADC 22 Mux' 'DMIC3'"
        cset "name='rt714 ADC 23 Mux' 'DMIC4'"
        cset "name='rt714 FU02 Capture Switch' 1"
        cset "name='rt714 FU02 Capture Volume' 47"
]

FixedBootSequence [
        exec "-/sbin/modprobe snd_ctl_led"
        sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:rt714 FU02 Capture Switch"
]
```

bardliao commented:
> @bardliao, you cannot change the name_prefix as UCM expects `rt714 ` for sdca variants:
> 
> ```
> # cat /usr/share/alsa/ucm2/codecs/rt715-sdca/init.conf
> # RT715-sdca (aka RT714) specific volume control settings
> 
> BootSequence [
>         cset "name='rt714 FU0A Capture Switch' 0"
>         cset "name='rt714 ADC 22 Mux' 'DMIC3'"
>         cset "name='rt714 ADC 23 Mux' 'DMIC4'"
>         cset "name='rt714 FU02 Capture Switch' 1"
>         cset "name='rt714 FU02 Capture Volume' 47"
> ]
> 
> FixedBootSequence [
>         exec "-/sbin/modprobe snd_ctl_led"
>         sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:rt714 FU02 Capture Switch"
> ]
> ```

Thanks @ujfalusi. I removed the changing name_prefix commit now. I think we can follow UCMs convention to use rt714 name_prefix for rt715-sdca codec.

ujfalusi commented:
@bardliao, can you mark these as fixes for 8266c73126b75eabbebefe7ce489a798e9ef2662 ?
I think that is the point when things got broken.

bardliao committed:
ASoC: Intel: sof_sdw: use generic rtd_init function for Realtek SDW
DMICs

The only thing that the rt_xxx_rtd_init() functions do is to set
card->components. And we can set card->components with name_prefix
as rt712_sdca_dmic_rtd_init() does.
And sof_sdw_rtd_init() will always select the first dai with the
given dai->name from codec_info_list[]. Unfortunately, we have
different codecs with the same dai name. For example, dai name of
rt715 and rt715-sdca are both "rt715-aif2". Using a generic rtd_init
allow sof_sdw_rtd_init() run the rtd_init() callback from a similar
codec dai.

Fixes: 8266c73126b7 ("ASoC: Intel: sof_sdw: add common sdw dai link init")
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw: remove unused rt dmic rtd_init

Some Realtek SDW DMIC codecs use the generic rt_dmic_rtd_init callback now.
Remove unused rtd_init callbacks.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt722_sdca: set rtd_init in codec_info_list[]

Add rt722 into rt_sdca_jack_rtd_init() supported list, and set
rtd_init callback directly in codec_info_list[]. No need to use a
init callback to set the rtd_init callback.
Besides, sof_sdw_rt_sdca_jack_init() doesn't set .rtd_init anymore.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw_rt722_sdca: use rt_dmic_rtd_init

rt_dmic_rtd_init() can be used for rt722 dmic, too.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
> @bardliao, can you mark these as fixes for [8266c73](https://github.com/thesofproject/linux/commit/8266c73126b75eabbebefe7ce489a798e9ef2662) ? I think that is the point when things got broken.

Done with reordering the commits. And it won't apply to https://github.com/thesofproject/linux/commit/8266c73126b75eabbebefe7ce489a798e9ef2662, but will apply to https://github.com/thesofproject/linux/commit/579d6596ebea488ad661bfa484c771c2b47eecc5 which is the next commit in the same series. I assume that is ok, right?

ujfalusi commented on sound/soc/intel/boards/sof_sdw_rt_dmic.c:
```diff
@@ -0,0 +1,52 @@
+	struct snd_soc_dai *codec_dai;
+	char *mic_name;
+
+	codec_dai = get_codec_dai_by_name(rtd, dmics, ARRAY_SIZE(dmics));
```
I guess it is not possible to have both rt715 and rt712, or a mix if rt715 and rt715-sdca?

bardliao replied:
> I guess it is not possible to have both rt715 and rt712, or a mix if rt715 and rt715-sdca?

No, we don't have such configuration. We have at most one SDW DMIC. But, if it really happens, `rt_dmic_rtd_init()` will be called twice, and both rt715 and rt712 will be appended to `card->components`.

bardliao replied:
> > I guess it is not possible to have both rt715 and rt712, or a mix if rt715 and rt715-sdca?
> 
> No, we don't have such configuration. We have at most one SDW DMIC. But, if it really happens, `rt_dmic_rtd_init()` will be called twice, and both rt715 and rt712 will be appended to `card->components`.

Sorry, I was wrong. `rt_dmic_rtd_init()` will not be called twice since the `rtd_init_done` will be set.

ujfalusi reviewed (commented):
@bardliao, while this fixes the regression on the laptop I have, I think it is still leaves at least one bug:
The sof_sdw_rtd_init() will update the wrong codec_info entry if non sdca rt715 is used on the board.
We just happen to have the same rtd_init() callback set for all dmic codecs which will set the correct component name.
Not only for the rt715/rt715-sdca/rt714, but rt712-sdca-aif1 and rt712-sdca-dmic-aif1 also have duplicate entry in the struct.

Is this something we should be concerned about?

bardliao commented:
> @bardliao, while this fixes the regression on the laptop I have, I think it is still leaves at least one bug: The sof_sdw_rtd_init() will update the wrong codec_info entry if non sdca rt715 is used on the board. We just happen to have the same rtd_init() callback set for all dmic codecs which will set the correct component name. Not only for the rt715/rt715-sdca/rt714, but rt712-sdca-aif1 and rt712-sdca-dmic-aif1 also have duplicate entry in the struct.
> 
> Is this something we should be concerned about?

Hmm, I think it should be fine. The duplicated dai name share the same codec driver except rt715 and rt715-sdca. And rt715 and rt715-sdca can use the same rtd_init.

plbossart commented on sound/soc/intel/boards/sof_sdw_rt_dmic.c:
```diff
@@ -15,6 +15,7 @@
 static const char * const dmics[] = {
 	"rt715",
 	"rt712-sdca-dmic",
+	"rt722-sdca",
```
can you add a comment on why we don't have rt715-sdca in the list? I am a bit lost on conventions.

bardliao replied:
> can you add a comment on why we don't have rt715-sdca in the list? I am a bit lost on conventions.

Because rt715-sdca codec driver uses [rt715-aif1](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/codecs/rt715-sdca.c#L936), too. I will add a comment on it. BTW, if rt715-sdca codec driver can change its dai name to `rt715-sdca-aif1` and `rt715-sdca-aif2`, it would help machine driver to distinguish rt715 and rt715-sdca DAIs. What do you think @shumingfan?

shumingfan replied:
@bardliao I think the rt715-sdca codec driver could change the dai name to `rt715-sdca-aif1` and `rt715-sdca-aif2`.
The rt711-sdca codec driver did the same thing.

bardliao committed:
ASoC: rt715-sdca: rename dai name with rt715-sdca prefix

The dai name are the same as rt715 codec. Rename them with rt715-sdca
prefix allow machine driver to distinguish rt715 and rt715-sdca from
dai name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: sof_sdw: change rt715-sdca dai name

As rt715-sdca dai name has been changed in codec driver, we should
change it in machine driver, too. Changing the dai name in
codec_info_list[] also help sof_sdw_rtd_init() to run .rtd_init()
from the same codec.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented:
@ujfalusi can you check if this PR fixes the issue #4834
